### ISSUE TITLE
How to configure private key?!

### ISSUE THREAD

**Issue by Psiiirus** [tapkey/tapkey-api-samples-java#5]
Hiya,
I read all the documentation but just don't get how to set up the private key for this application.
I setted up all the server key things online and filled out the application.properties but /public-key gives me a different public key. I also wondering why the documentation don't use the passphrase in the java example (tutorial and this repo).

@moritzgloeckl  @sebastianhaas  I hope some one has a spare minute to give me a clue :-)

**Comment by sebastianhaas** [comment]
The key pair is created every time the application starts and held only in-memory. This is the relevant line: https://github.com/tapkey/tapkey-api-samples-java/blob/master/token-exchange-server-sample/src/main/java/io/tapkey/developers/service/SampleKeyService.java#L14

As you can see, the public key printed on `/public-key` is retrieved from the key service initialized in SampleKeyService.java#L14 before:
https://github.com/tapkey/tapkey-api-samples-java/blob/master/token-exchange-server-sample/src/main/java/io/tapkey/developers/controllers/IdentityController.java#L48

That probably answers your third question too: The example server in this repository creates key pairs on the fly, hence there is no way to use the matching public key in the examples on developers.tapkey.io (https://developers.tapkey.io/api/authentication/identity_providers/).

**Comment by Psiiirus** [comment]
Hi @sebastianhaas minutes ago i noticed the /public-key is changing every restart! Makes sense so for testing purpose i will just du a key rollover every time i restart the token exchange server.
____________________________________________________________
The problem is it still doesn't work...the ios-sample creates the user but the logs gives me:
[e55397ec] 500 Server Error for HTTP GET "/user/tapkey-token"
and
403 Forbidden from PUT https://my.tapkey.com/api/v1/Owners/<removed>/IdentityProviders/<removed>/Users

Just to be sure i've create an 
- Identity Provider to be the interface between tapkey users and my user management
- Token Exchange so my user can get access to the logs
- Client Credentials so the java sample can interact with the tapkey oauth

did i get this right?

**Comment by sebastianhaas** [comment]
@Psiiirus yes, that is the intended behaviour. Please feel free to create a PR for this repository's readme and add a few lines that you would have found helpful yourself. Yes, you will need to perform a key rollover on Tapkey side every time you restart this sample server. Alternatively, you can replace the line I've mentioned above with a persistent key storage.

Now re your actual problem:
From what you've written, that seems all good and correct. You will need to provide more details I guess, e.g. include the requests and reponses in a proper timeline, also including the respective authorization headers.

What makes me wonder, you state that ".the ios-sample creates the user" but  then you state that "403 Forbidden from PUT https://my.tapkey.com/api/v1/Owners//IdentityProviders//Users" which is exactly the operation that would actually create the user on Tapkey side but clearly fails due to missing authorization. Have you added `write:ip:users` (and read) to your client credentials client?

**Comment by minichma** [comment]
@sebastianhaas Thanks again! :-)

**Comment by minichma** [comment]
The problem probably is that the Identity Provider Id you specified is wrong. For the client credentials user you need to specify `tkIdp`.

**Comment by Psiiirus** [comment]
@minichma and the `tkIdp` is not the Identity Provider Id of "Local Java Server" ?

**Comment by Psiiirus** [comment]
omg i see. thanks @L6Echo

**Comment by L6Echo** [comment]
Thank you, for your feedback. We are going to improve the README.
